### ISSUE TITLE
Accuracy cannot be guaranteed when training on Linux server

### ISSUE THREAD

**Issue by Rhierarch** [ultralytics/ultralytics#6866]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

When I trained my datasets with yolov8n.pt on Linux server,I found the loss is always bigger than trained on windows and the accuracy cannot be guaranteed.

### Train process on windows:
![results](https://github.com/ultralytics/ultralytics/assets/128219842/8f7851d8-7c8e-4afc-a6da-9636749445ee)

### Train process on Linux server:
![results (1)](https://github.com/ultralytics/ultralytics/assets/128219842/d1f02b8b-c39c-4284-8d1a-0f37668c170f)

When I try to use the best.pt to detect new images, the detection is basically successful with model trained on windows. However,it can not detect anything with model trained on Linux!

I have compared args.yaml and datasets on both  windows and Linux, and found that they are all the same.

Is that caused by the difference between windows and Linux or the difference between the environment of them?
But the strange thing is that the detection part works well on both windows and Linux server.
 

### Environment

Linux environment:
system:Linux version 3.10.0-1160.el7.x86_64 (root@centos-build-7.9) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-44) (GCC) ) 
#1 SMP Wed Nov 18 03:43:48 UTC 2020

YOLOv8.0.221 🚀 Python-3.10.13 torch-2.0.1+cu117 CUDA:0 (NVIDIA A40, 45494MiB)
cuda=11.7
torch                 2.0.1+cu117
torchaudio            2.0.2+cu117
torchvision           0.15.2+cu117
ultralytics           8.0.221
numba                 0.58.0
numpy                 1.24.1

The task is allocated resources by the PBS scheduling system on Linux server.

Windows environment:
system: Microsoft Windows [Version 10.0.19042.928]

YOLOv8.0.221 🚀 Python=3.10.2 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4060, 8188MiB)
cuda=12.3
torch                        2.1.0+cu121
torchaudio                   2.1.0+cu121
torchvision                  0.16.0+cu121
numba                        0.57.1
numpy                        1.23.4

### Minimal Reproducible Example

python script on Linux server:

from ultralytics import YOLO
model = YOLO('yolov8n,pt') 
results = model.train(name='result,
                      data='mydata.yaml',
                     epochs=100,
                      imgsz=640,
                      workers=32,
                      batch=8
                      device=0,
                      )

python script on Windows:

from ultralytics import YOLO
model = YOLO('yolov8n,pt') 
results = model.train(name='result,
                      data='mydata.yaml',
                     epochs=50,
                      imgsz=640,
                      workers=0,
                      batch=8
                      device=0,
                      )

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Rhierarch hello! Thank you for the detailed report. Variability in training results between different operating systems can occur due to numerous factors. The core difference often lies in how the underlying libraries and drivers interact with the OS and hardware, especially when dealing with different versions of CUDA, PyTorch, or even the graphics drivers. 

This variability can occasionally translate into different training outcomes despite having the same hyperparameters and dataset. In particular, CUDA version mismatches and differences in the PyTorch or TorchVision versions between environments can lead to subtle differences in operations that may affect reproducibility. Also, the number of workers set during training might influence the data loading process and results.

Given that the detection part works well across the environments suggests that the software is functioning correctly. I would recommend ensuring that both systems are using the same versions of dependencies as much as possible. Additionally, consider controlling for batch size and number of workers, as these can impact the results.

The difference in devices (NVIDIA A40 vs GeForce RTX 4060) and the associated system resources could also be a factor. Investigating potential CUDNN flags or PyTorch settings for deterministic operations could help further.

For more details on how to address this, consult our documentation that discusses best practices for reproducibility and troubleshooting. If you continue to experience issues, please provide any further details, and we'll be happy to assist further.

**Comment by Rhierarch** [comment]
@glenn-jocher 
Thank you for your answer！
I thought the problem was caused by batch size and number of workers or other configurations at first, so I tried several combinations of configurations on platforms mentioned above, but their impact wasn't obvious. As long as the platform is the same, changing the configuration doesn't have much impact on both convergence and metrics.

And when I took a closer look at the differences in the training process between this two platforms, I noticed that their loss was just different in the beginning. I guess the reason is that they have different loss functions or different ways of calculating them.

So I haved try the same task on another Windows platform without GPU and cuda, and it get the same result as my NVIDIA GeForce RTX 4060 platform mentioned in my report.

The results of these tests made it difficult for me to determine whether the problem was due to a difference in the environment or the OS. 

Unfortunately, I was not able to modify the GPU driver and cuda versions of the linux server, which were pre-installed by administrators, and I was not able to modify the above versions with administrator privileges. But I can try to talk to my administrator later and ask him to upgrade the GPU driver and cuda version, and then try again.  

Before that, in addition to try to downgrade my windows platform cuda to the same version as linux and try the effect, I have no better way to determine the source of the problem, I hope you can give me some suggestions!

**Comment by Burhan-Q** [comment]
@Rhierarch instead of downgrading CUDA, you could match the PyTorch versions installed for both instances. It looks like you have `torch2.1.0` and `torch 2.0.1` installs. Additionally, I'd recommend using the `deterministic` argument and the same seed values when testing your training. This should help (not ensure) closer results. Finally, if you use a public dataset like `coco128` (it's a toy dataset but might help find glaring discrepancies) or `VisDrone` (sizeable but not gigantic) dataset, then others could try to replicate your test config. If you want to test a short training run with the VisDrone dataset, let's say 50 epochs using a nano model with deterministic seed, that might be a quick way to verify performance.

I will share that I've trained models on my Windows 10 PC and an Ubuntu PC, and have not noticed any glaring differences in outcomes. These were trainings that were used for testing other functions/features and were not going to be deployed, so I wasn't paying _very close_ attention to the loss profile or reviewing all class metrics.

**Comment by Rhierarch** [comment]
@Burhan-Q  Thanks for your suggestions! I tried to train coco128 datasets on my Windows 10 PC  and  Linux in 50 epochs, here is the results :
### Train process on Windows 10:
![results](https://github.com/ultralytics/ultralytics/assets/128219842/36c12f4c-f815-4484-a2c7-815eddc99367)

### Train process on Linux server:
![results](https://github.com/ultralytics/ultralytics/assets/128219842/753c89a0-dff2-434b-86f9-ccda8cfef63a)


It seems that thet have not obvious differences in outcomes between two platforms, so it is  possible that the difference is due to my custom dataset.

What I had trouble understanding was that training on windows worked well, and that the accuracy of yolov8 training on my dataset made such a big difference when I had trained on other tasks on both platforms. Is it because my dataset is not well prepared and there is a chance that it will not converge, but the linux server and its environment magnifies this possibility?

**Comment by Burhan-Q** [comment]
@Rhierarch I suspect there could be multiple factors at play here. On my Windows 10 and Linux PCs, I quickly ran a few trainings for the COCO128 dataset. 

Win10 PC
```
Intel 10th gen CPU
RTX3080 12 GB GPU
Ultralytics YOLOv8.0.223
Python-3.12.0
torch-2.2.0.dev20231205+cu121
# NOTE this configuration is not officially supported but was the last active venv
```

Linux PC
```
Ryzen 5 1600 CPU
RTX2060 6GB GPU
Ultralytics YOLOv8.0.215
Python-3.10.12
torch-2.1.0+cu121
```

## Win10 Train-1 (default workers)

`yolo mode=train task=detect model=yolov8n.pt data=Q:\datasets\coco128\coco128.yaml epochs=50`

<details>

<summary>Expand for results plots</summary>

![results](https://github.com/ultralytics/ultralytics/assets/62214284/99d00a6b-e937-49f0-984f-9207293e4c79)

</details>

## Win10 Train-2 (workers=0)

`yolo mode=train task=detect model=yolov8n.pt data=Q:\datasets\coco128\coco128.yaml epochs=50 workers=0`

<details>

<summary>Expand for results plots</summary>

![results](https://github.com/ultralytics/ultralytics/assets/62214284/e99546b7-75f9-441c-aa13-86d8f9dffbb1)

</details>

## Win10 Train-3 (default workers + deterministic)

`yolo mode=train task=detect model=yolov8n.pt data=Q:\datasets\coco128\coco128.yaml epochs=50 deterministic=True seed=7`

<details>

<summary>Expand for results plots</summary>

![results](https://github.com/ultralytics/ultralytics/assets/62214284/1974f6ed-6070-4956-b2bd-46708c074aed)

</details>

## Win10 Train-4 (workers=0 + deterministic)

`yolo mode=train task=detect model=yolov8n.pt data=Q:\datasets\coco128\coco128.yaml epochs=50 deterministic=True seed=7 workers=0`

<details>

<summary>Expand for results plots</summary>

![results](https://github.com/ultralytics/ultralytics/assets/62214284/344d0f54-5ccf-4a87-a741-2fd3fc3b4437)

</details>

---

As you can see, changing the number of `workers` will have an influence on the outcome even on the same machine and environment, but off-hand I'm not knowledgeable as to why that would be the case. I suspect that minor differences in the system processor threads/cores could slightly nudge float values in one direction or another. Given your initial commands are `workers=0` and `workers=32` this could stack to make for a big difference

Switching to my Linux PC, you can see that the number of workers will also influence the outcome as well. This machine not only has a different OS, but also a completely different hardware than my Win10 PC and uses `Ultralytics YOLOv8.0.215 Python-3.10.12 torch-2.1.0+cu121`

## Linux Train-1 (default workers)

`yolo mode=train task=detect data=datasets/coco128.yaml epochs=50 workers=0`

<details>

<summary>Expand for results plots</summary>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/2b4dec80-3c73-4b45-8e25-9bdb2472237c)

</details>

## Linux Train-2 (workers=0)

`yolo mode=train task=detect data=datasets/coco128.yaml epochs=50 workers=0`

<details>

<summary>Expand for results plots</summary>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/0f4c05cd-76b4-4ac8-80e3-13f8582e3946)

</details>

---

The biggest factors that will likely impact changes in training outcomes (no particular order):

- number of workers
- hardware architecture differences
- PyTorch version
- Ultralytics version
- python version
- other training args (hyperparameters, image size, batch size, deterministic, etc.)

**Comment by Burhan-Q** [comment]
@Rhierarch I also see that from your original post/results that you are using different values for `epochs` this will absolutely have an effect from the start of training. 

> python script on Linux server:
> 
> from ultralytics import YOLO
> model = YOLO('yolov8n,pt')
> results = model.train(name='result,
> data='mydata.yaml',
> epochs=100, # <- setting is different
> imgsz=640,
> workers=32, # <- setting is different
> batch=8
> device=0,
> )
> 
> python script on Windows:
> 
> from ultralytics import YOLO
> model = YOLO('yolov8n,pt')
> results = model.train(name='result,
> data='mydata.yaml',
> epochs=50, # <- setting is different
> imgsz=640,
> workers=0, # <- setting is different
> batch=8
> device=0,
> )

For a true equal comparison, you will need to ensure parameters such as `epochs`, `batch`, `imgsz`, and `workers` are configured to be the same on both systems. Using the default configuration where possible would make for the easiest cross-platform comparison. Although not explicitly shown, in your [results plots](https://github.com/ultralytics/ultralytics/issues/6866#issuecomment-1847086895) from testing the COCO128 dataset the number of epochs appear to be the same and so the starting losses are much closer. If you're training with starting weights, it would probably be helpful to ensure these are the _exact_ same starting weights file as well. 

Beyond ensuring your configurations for training match exactly, my prior recommendation for matching versions of PyTorch between systems would be the next step that I would take to achieve parity between these systems. With respect to predictions, if you test with pretrained model weights on a subset or toy dataset, I suspect that you will find little if any difference in detection performance.

**Comment by Rhierarch** [comment]
@glenn-jocher Another thing to note is that I noticed a difference in the accuracy of the location coordinates of the labels in my dataset compared to the coco128. This difference is caused by the labeltool I used. Maybe I should keep them at the same precision and compare them

### coco128 label file example
![微信图片_20231208234837](https://github.com/ultralytics/ultralytics/assets/128219842/df1acbf8-7ba8-41c8-9963-4ea3b03bc104)


### my label file example
![微信截图_20231208234807](https://github.com/ultralytics/ultralytics/assets/128219842/323bf2ca-8c2b-4006-b563-5a59ec37f442)

**Comment by Burhan-Q** [comment]
Truncating the bounding box coordinates will certainly help ensure that any floating point differences don't stack up too significantly.

**Comment by Rhierarch** [comment]
@glenn-jocher When I followed the coco dataset and modified the labeling tool I used to truncate the floating-point numbers with six significant digits in the box coordinates , the resulting dataset was trained on a Linux server and finally achieved the same level of performance as windows!

### Train process on Linux server after truncating box coordinate:
![results (1)](https://github.com/ultralytics/ultralytics/assets/128219842/47a67a93-2f52-4fb3-bedd-ef013f719dc2)


I think it's because the Red Hat system on the Linux server reads floating point digits differently than windows. By the way, I did not achieve this by setting FP16=True in the training parameter. Maybe it just changes the model parameters and input image accuracy without truncating the input label data.

All in all, this suggests that having too many floating-point digits in the labeled data may lead to error amplification and even failure of the model to converge. And this is more likely to happen on linux systems.

**Comment by glenn-jocher** [comment]
@Rhierarch glad to hear that adjusting the precision of box coordinates resolved the discrepancies in your training outcomes on different operating systems. Floating-point representation can indeed behave differently across platforms, and it's insightful that truncating to six significant digits aligned the performance. This could certainly be a valuable consideration for others encountering similar cross-platform training variances. Thank you for sharing your findings with the community! 🌟
